### ISSUE TITLE
Compiler Bug: Could not find type of <Declaration_Variable> for declaration of same name

### ISSUE THREAD

**Issue by fruffy** [p4lang/p4c#2544]
Simply declaring a new local header with the same name leads to a compiler bug currently.

```p4
control ingress(inout Headers h) {
    apply {
        Headers h = h;
    }
}
```

Output:

```
Deprecated_1_CheckDeprecated
FrontEnd_9_Deprecated
FrontEnd_10_CheckNamedArgs
In file: p4c/frontends/p4/typeChecking/typeChecker.cpp:168
[31mCompiler Bug[0m: bugs/crash/same_name_crash.p4(24): Could not find type of <Declaration_Variable>(238) h/34 Headers h = h;
        Headers h = h;
        ^^^^^^^^^^^^^^*/
```
[same_name_crash.p4.txt](https://github.com/p4lang/p4c/files/5217691/same_name_crash.p4.txt)

**Comment by mbudiu-vmw** [comment]
@ChrisDodd : this looks like a bug in the resolveReferences pass; the h on the right-hand side is resolved as the local variable instead of being resolved as the parameter. I believe that this is due to your recent refactoring of this pass. The variable declaration does introduce a new symbol, but it should be only visible after the declaration.

**Comment by dbrajovic-syrmia** [comment]
@ChrisDodd : If possible, I would like to assign this issue to myself for further investigation - I'm interested in contributing to the project and the issue seems fit for a starting point.

**Comment by mbudiu-vmw** [comment]
If you submit a PR and it's good then we'll accept it. I have added you to the list of assignees.

**Comment by mbudiu-vmw** [comment]
Fixed by #2589
